src/Knp/DictionaryBundle/Dictionary/CallableDictionary.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+namespace Knp\DictionaryBundle\Dictionary;
+
+use Knp\DictionaryBundle\Dictionary as DictionaryInterface;
+
+class CallableDictionary implements DictionaryInterface
+{
+    /**
+     * @var string
+     */
+    private $name;
+
+    /**
+     * @var mixed[]|\ArrayAccess
+     */
+    private $values = null;
+
+    /**
+     * @var callable
+     */
+    private $callable;
+
+    /**
+     * @param string   $name
+     * @param callable $callable
+     */
+    public function __construct($name, $callable)
+    {
+        $this->name = $name;
+
+        if (false === is_callable($callable)) {
+            throw new \InvalidArgumentException('Second argument must be a callable.');
+        }
+
+        $this->callable = $callable;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getValues()
+    {
+        $this->hydrate();
+
+        return $this->values;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getKeys()
+    {
+        $this->hydrate();
+
+        return array_keys($this->values);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function offsetExists($offset)
+    {
+        $this->hydrate();
+
+        return array_key_exists($offset, $this->values);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return mixed
+     */
+    public function offsetGet($offset)
+    {
+        $this->hydrate();
+
+        return $this->values[$offset];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function offsetSet($offset, $value)
+    {
+        $this->hydrate();
+
+        $this->values[$offset] = $value;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function offsetUnset($offset)
+    {
+        $this->hydrate();
+
+        unset($this->values[$offset]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getIterator()
+    {
+        $this->hydrate();
+
+        return new \ArrayIterator($this->values);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function serialize()
+    {
+        $this->hydrate();
+
+        return serialize(array(
+            'name'   => $this->name,
+            'values' => $this->values,
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function unserialize($serialized)
+    {
+        $data = unserialize($serialized);
+
+        $this->name   = $data['name'];
+        $this->values = $data['values'];
+    }
+
+    /**
+     * Hydrate values from callable.
+     */
+    protected function hydrate()
```

Comment:
If this method is protected, I think `values` and `callable` should be protected too.